tactics/equality.ml

```diff
@@ -311,29 +311,145 @@ let general_elim_clause with_evars frzevars tac cls c (ctx, eqn, args) l l2r eli
-  let open UnivGen.QualityOrSet in
-  let name =
-    if l2r then
-      match sort with
-      | Qual (QConstant QProp) -> "core.eq.ind_r"
-      | Qual (QConstant QSProp) -> "core.eq.sind_r"
-      | Set | Qual (QConstant QType | QVar _) -> "core.eq.rect_r"
-    else
-      match sort with
-      | Qual (QConstant QProp) -> "core.eq.ind"
-      | Qual (QConstant QSProp) -> "core.eq.sind"
-      | Set | Qual (QConstant QType | QVar _) -> "core.eq.rect"
-  in
-  Rocqlib.lib_ref_opt name
+(* eq_scheme_name returns the elimination principle to be used when dealing with eq, dependending on dep(endency), in conclusion or not and left-to-right *)
+(* this could be handled by has_J_ref as well, but is more efficient has it bypasses the typeclass search *)
+let eq_scheme_name dep lft2rgt inccl target = let open Sorts.Quality in
+  match dep, lft2rgt, inccl, target with
+    (* Non dependent case *)
+    | false, Some true, true , QConstant QType -> Rocqlib.lib_ref_opt "core.eq.rect_r" , 5
+    | false, Some true, true , QConstant QProp -> Rocqlib.lib_ref_opt "core.eq.ind_r" , 5
+    | false, Some true, false , QConstant QType -> Rocqlib.lib_ref_opt "core.eq.rect" , 5
+    | false, Some true, false , QConstant QProp -> Rocqlib.lib_ref_opt "core.eq.ind" , 5
+    | false, _ , false , QConstant QType -> Rocqlib.lib_ref_opt "core.eq.rect_r" , 5
+    | false, _ , false , QConstant QProp -> Rocqlib.lib_ref_opt "core.eq.ind_r" , 5
+    | false, _ , true , QConstant QProp -> Rocqlib.lib_ref_opt "core.eq.ind" , 5
+    | false, _ , true , QConstant QType -> Rocqlib.lib_ref_opt "core.eq.rect" , 5
+    (* Dependent case *)
+    | true, Some true, true , QConstant QType  -> Rocqlib.lib_ref_opt "core.eq.rect_r_dep", 5
+    | true, Some true, true , QConstant QProp  -> Rocqlib.lib_ref_opt "core.eq.ind_r_dep" , 5
+    | true, _ , true , QConstant QType  -> Rocqlib.lib_ref_opt "core.eq.rect_dep" , 5
+    | true, _ , true , QConstant QProp  -> Rocqlib.lib_ref_opt "core.eq.ind_dep" , 5
+    | _ , _, _ , _ -> None , 0
+
+
+(* has_J_ref returns the name of the class to be used, dependending on dep(endency), in conclusion or not and left-to-right *)
+(* The integer encodes the position of the equality argument in the elimination principle, starting from 0 *)
+let has_J_ref dep lft2rgt inccl =
+  match dep, lft2rgt, inccl with
+    (* Non dependent case *)
+    | false, Some true, true -> Rocqlib.lib_ref "rocq.core.Has_Leibniz_r" , 5
+    | false, Some true, false -> Rocqlib.lib_ref "rocq.core.Has_Leibniz" , 5
+    | false, _, false -> Rocqlib.lib_ref "rocq.core.Has_Leibniz_r" , 5
+    | false, _, true -> Rocqlib.lib_ref "rocq.core.Has_Leibniz" , 5
+    (* Dependent case *)
+    | true, Some true, true -> Rocqlib.lib_ref "rocq.core.Has_J_r" , 5
+    | true, Some true, false -> Rocqlib.lib_ref "rocq.core.Has_J_r_forward" , 4
+    | true, _, true -> Rocqlib.lib_ref "rocq.core.Has_J" , 5
+    | true, _, false -> Rocqlib.lib_ref "rocq.core.Has_J_forward" , 4
```

Comment:
Why are these registers `rocq.core.*` while all other definitions from the corelib are `core.*`?